src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/SearchParameters/SearchParameterFixtureData.cs

```diff
@@ -23,25 +24,35 @@ namespace Microsoft.Health.Fhir.Core.UnitTests.Features.Search
 {
     public class SearchParameterFixtureData
     {
+        // this type is immutable and is safe to reuse
+        private static FhirNodeToSearchValueTypeConverterManager _fhirNodeToSearchValueTypeConverterManager;
+
+        private SearchParameterDefinitionManager _searchDefinitionManager;
+        private SupportedSearchParameterDefinitionManager _supportedSearchDefinitionManager;
+
         static SearchParameterFixtureData()
         {
             FhirPathCompiler.DefaultSymbolTable.AddFhirExtensions();
-
-            Compiler = new FhirPathCompiler();
-            Manager = CreateFhirElementToSearchValueTypeConverterManager();
-            SearchDefinitionManager = CreateSearchParameterDefinitionManager(new VersionSpecificModelInfoProvider());
-            SupportedSearchDefinitionManager = new SupportedSearchParameterDefinitionManager(SearchDefinitionManager);
         }
 
-        public static SearchParameterDefinitionManager SearchDefinitionManager { get; }
+        public static FhirPathCompiler Compiler { get; } = new FhirPathCompiler();
 
-        public static SupportedSearchParameterDefinitionManager SupportedSearchDefinitionManager { get; }
+        public async Task<SearchParameterDefinitionManager> GetSearchDefinitionManager()
+        {
+            return _searchDefinitionManager ??= await CreateSearchParameterDefinitionManager(new VersionSpecificModelInfoProvider());
+        }
 
-        public static FhirNodeToSearchValueTypeConverterManager Manager { get; } = CreateFhirElementToSearchValueTypeConverterManager();
+        public async Task<SupportedSearchParameterDefinitionManager> GetSupportedSearchDefinitionManager()
+        {
+            return _supportedSearchDefinitionManager ??= new SupportedSearchParameterDefinitionManager(await GetSearchDefinitionManager());
```

Comment:
>GetSearchDefinitionManager [](start = 109, length = 26)

Changes here happened mostly because it's impossible to have async constructor.
I understand what as soon as you go async, everything upstream has to be async as well. But if downstream code you using remain sync, you can continue to be sync in upstream until you reach async behavior.

I've tried to track classes, and they mostly look like they remain sync, just wrapped in async logic. (can be easily wrong and maybe there is dependency on async code somewhere at the bottom)
Is it possible to have two interfaces and have `IAsyncHostedService` and `IHostedService` and use them appropriately?

Yes, this would make people question which interface to use, and maybe eventually everything would converge to async version, but on plus side, some code can be simplified, like here, or other places where all is changed is `async` added with Task.Complete.

Reply:
We had `IStartable` that was synchronous, this PR was to move everything to IHostedService which provides the Async Start method. I am interested to learn your experience here, what do you see as the biggest downsides/concerns for having only the async version? Is it the sync=>async wrapping logic/mismatch, or performance or other?

Reply:
Never run benchmarks to compare how sync vs async code would perform (if both act in a sync way) (I run tonight, for my own education), but I doubt performance wise it's stark contrast.

My main concern is mental overload which comes with async. Why be complex if you can be simple.

That been said, I just notice GetAwaiter() in line 79 (left) which mean this whole thing need to be async anyway.

---
In reply to: [512256266](https://github.com/microsoft/fhir-server/pull/1396#discussion_r512256266) [](ancestors = 512256266)